### PR TITLE
Feature/aruco speedup

### PR DESCRIPTION
**Pull request by urbste**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

Hi,

I improved the Aruco module (I will call it Aruco3 as the authors did). It implements the improvements that were proposed in the paper: "Speeded up detection of squared fiducial markers", 2018, FJ Romera-Ramirez et al.

The implementation is not entirely finished, meaning that I would like to know from you guys where I should add tests and so on. 
I uploaded simple test videos: https://drive.google.com/drive/folders/153UTdWMOwxIsn2nxdWW0itV8na4AANZH?usp=sharing
The folder contains 3 videos with 4k, 1k and VGA resolution to test the speed improvements and detection rates. However are there any particular tests for the Aruco module to compare against the original implementation?

I added the functionality to the example_aruco_detect_markers.cpp file. In addition, I added new flags to the detector_params.yml. If Aruco3 is turned off, the code should behave as the original Aruco implementation.
```yaml
# new aruco functionality
useAruco3Detection: 0
minSideLengthCanonicalImg: 32 # 16, 32, 64 --> tau_c from the paper
minMarkerLengthRatioOriginalImg: 0.08 # range [0,0.2] --> tau_i from the paper
cameraMotionSpeed: 0.5 # range [0,1) --> tau_s from the paper
useGlobalThreshold: 0
```
The parameters are explained in the documentation and reference to equations and sections from the paper.

Some preliminary results for #detected and mean computation per image in [ms]:
Call: 
``` bash
./example_aruco_detect_markers -v=4k_test.mp4 -d=16 -dp=detector_params.yml -refine=1
```
For all datasets: tau_c = 32pixel
| Dataset | Aruco | Aruco3 (tau_i = 0.1)  | Aruco3 (tau_i = 0.05)  | Aruco3 (tau_i = 0.02)  |
|--|--|--|--|--|
| VGA | 921 / 2.91ms |  704 / 0.95ms |  761 / 1.1ms  | 909 / 2.1ms |
| 1080p |  1186 / 14ms |  995 / 2.6ms| 1100 / 3.1ms | 1180 / 5.3ms |
| 4K | 1013 / 63ms | 931 / 7.6ms |  1011 / 8.1ms  |  1071 / 13.3ms | 

For all datasets: tau_c = 16pixel
| Dataset | Aruco | Aruco3 (tau_i = 0.1)  | Aruco3 (tau_i = 0.05)  | Aruco3 (tau_i = 0.02)  |
|--|--|--|--|--|
| VGA |  921 / 2.91ms  | 474 / 0.8ms | 796 / 0.9ms |  919 / 1.7ms |
| 1080p | 1186 / 14ms  | 132 / 2.9ms  | 1086 / 2.9 ms  | 1173 / 4.0 ms | 
| 4K | 1013 / 63ms | 118 / 7.0ms |  1013 /7.4ms  |  1052 / 8.3ms | 

Now instead of fixing tau_i we can also let the algorithm decide it by using the marker with the smallest perimeter from the last frame. This only works for video streams of course. In addition, we can use global thresholding instead of adaptive thresholding, assuming that the illuminations does not change much. 
Call: 
``` bash
./example_aruco_detect_markers -v=4k_test.mp4 -d=16 -dp=detector_params.yml -refine=1 -ar3vid
```

For all datasets: tau_c = 32pixel, tau_s = 0.1, useGlobalThreshold=1
| Dataset | Aruco | Aruco3 |
|--|--|--|
| VGA |  921 / 2.91ms  | 758 / 1.0ms  |
| 1080p | 1186 / 14ms  | 1174 / 2.9ms  | 
| 4K | 1013 / 63ms | 1036 / 8.8ms | 

For all datasets: tau_c = 16pixel, tau_s = 0.1, useGlobalThreshold=1
| Dataset | Aruco | Aruco3 |
|--|--|--|
| VGA |  921 / 2.91ms  | 905 / 0.7ms  |
| 1080p | 1186 / 14ms  | 1156 / 2.8ms  | 
| 4K | 1013 / 63ms | 1022 / 7.7ms | 

Let me know what you think.
Cheers
Steffen

### CODE REVIEW

**Comment by urbste** [comment]
Ok I can see that already some Aruco tests are failing. I will have a look into it.

**Comment by urbste** [comment]
I added tests. I also fixed some small bugs. Can someone check why the build checks are failing? When I click on details it looks like all tests run successfully.

**Comment by alalek** [comment]
You can see compilation warnings [here](http://pullrequest.opencv.org/#/summary/contrib) (should be resolved before merge).

File: modules/aruco/include/opencv2/aruco.hpp [hunk 1]
```diff
@@ -146,6 +146,33 @@ enum CornerRefineMethod{
...
+ *   in Section 3.7. and is particularly useful for video sequences.
+ *   The parameter tau_i has a direct influence on the processing speed. Instead of setting a fixed value for it,
+ *   it can be adjusted at the end of each frame using
+ *   tau_i = (1-tau_s)*P(v_s)/4 (eq. 6 in paper).
+ *   Where P(v_s) is the perimeter of the smallest marker that was detected in the last frame.
+ * - useGlobalThreshold: if we process a video, the assumption is, that the illumination conditions remains
+ *   constant and global instead of adaptive thresholding can be applied, speeding up the binarization step.
+ * - foundGlobalThreshold: internal variable. It is used to cache the variable to the next detector call.
+ * - otsuGlobalThreshold: internal variable. It is used to cache the global otsu threshold to the next detector call.
+ * - foundMarkerInLastFrames: internal variable. It is used to cache if markers were found in the last frame.
```
Comment (alalek) [inline modules/aruco/include/opencv2/aruco.hpp:175]: > internal variable

Parameters structs are used in OpenCV as non-modified inputs only.
This requirement is necessary for Python and other bindings.

For such video processing case with internal state it make sense to create dedicated API with detector class which store this internal state (use PImpl pattern, see `QRCodeDetector` class from objdetect module)

File: modules/aruco/src/aruco.cpp [hunk 2]
```diff
@@ -750,9 +847,9 @@ static Point3f _interpolate2Dline(const std::vector<cv::Point2f>& nContours){
  * @return Crossed Point
  */
 static Point2f _getCrossPoint(Point3f nLine1, Point3f nLine2){
-	Matx22f A(nLine1.x, nLine1.y, nLine2.x, nLine2.y);
-	Vec2f B(-nLine1.z, -nLine2.z);
-	return Vec2f(A.solve(B).val);
+    Matx22f A(nLine1.x, nLine1.y, nLine2.x, nLine2.y);
+    Vec2f B(-nLine1.z, -nLine2.z);
+    return Vec2f(A.solve(B).val);
```
Comment (alalek) [inline modules/aruco/src/aruco.cpp:852]: Please avoid such massive space-only changes.
It is really hard to see/track the significant changes of PRs.

Consider moving these "whitespace" changes to the separate first commit or separate PR.

**Review by alalek — COMMENTED** [review]
Thank you for the contribution!

Please take a look on comments below.

**Comment by asenyaev** [comment]
jenkins cn please retry a build

**Comment by asenyaev** [comment]
jenkins cn please retry a build

**Comment by joellyons** [comment]
What's the latest on this PR?  Has it been abandoned?

**Comment by urbste** [comment]
I am still having it in the back of my head. However, the requested changes seem to be quite significant and I did not find the time to look at it. We were using this branch in prototype, however, we moved away from it. So if someone likes to take over please go ahead!

**Comment by joellyons** [comment]
Thank you for the update. If you don't mind a bit more into... why did you move away from it? Were you having problems with it?

**Comment by urbste** [comment]
@joellyons no we did not have problems with it. We wanted to use it in WebAssembly running on a smartphone to track the pose in realtime. However our board had a lot of markers and we never reached >30fps on Android devices. So we kind of changed our algorithm and now only occasionally update the pose using the ARuco board. The Aruco3 algorithm however really shines if you want to track the board over time, so you can update the tau_i parameter and use the global otsu thesholding. So we decided to stick with the classic Aruco implementation.

Still with my PR it is possible to use the classic impementation. I will see if I find the time to finish this PR :/

**Comment by joellyons** [comment]
@urbste thank you so much for taking the time to explain. We are wanting to track a small board over time, so this sounds perfect. But, I agree it's hard to find extra time to work on open source projects like this.

File: modules/aruco/src/aruco.cpp [hunk 3]
```diff
@@ -362,61 +377,88 @@ static void _detectInitialCandidates(const Mat &grey, vector< vector< Point2f >
...
-    Mat image = _image.getMat();
-    CV_Assert(image.total() != 0);
+    Mat grey = _image.getMat();
+    CV_Assert(grey.total() != 0);
 
     /// 1. CONVERT TO GRAY
-    Mat grey;
-    _convertToGrey(image, grey);
+    //Mat grey;
+    //_convertToGrey(image, grey);
```
Comment (AleksandrPanov) [inline modules/aruco/src/aruco.cpp:447]: ?
